### PR TITLE
Combobox examples with listbox popup: Add properties and states to listbox control button

### LINKED ISSUE #1333: Combobox examples: Mark up open button so it reflects combobox state
When using a touch based screen reader or using the reading cursor of a desktop screen reader, the open button of the combobox examples does not communicate the state of the combobox. Also the accessible "name of "Open" is insufficient.

Make the following changes so the open button is accessible:
1. Change the name to match the name of the element that the button controls. For example, in the comboboxes for choosing a "State", the name of the button should be "States".
2. Add aria-expanded to the button and set it to "true" when the combobox is expanded and "false when it is collapsed.
3. Add aria-controls to the button referencing the element that it controls.

### PR DESCRIPTION
**Pull request by jongund**
Fixes #1333 and #1331:
* Updates button to include aria attributes identified in issue #1333 and issue #1331  
* Adds regression test for the newly added attributes.

#### Preview link

[View the combobox examples in the compare branch with RawGitHack](https://raw.githack.com/w3c/aria-practices/issue1333-add-aria-to-button/examples/combobox/combobox-autocomplete-both.html)

### Review checklist

* [x] [Editorial review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#editorial_review) by Matt (@mcking65)
* [x] [Functional review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#functional_review) by Matt (@mcking65)
* [x] [Visual design review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#design_review) by Simon (@zcorpan)
* [x] [Accessibility review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#accessibility_review) by Matt (@mcking65)
* [x] [Code review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#code_review) by Simon (@zcorpan)
* [x] [Test review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#test_review) by Simon (@zcorpan)

### CODE REVIEW

File: examples/combobox/combobox-autocomplete-both.html [hunk 1]
```diff
@@ -51,16 +51,16 @@ <h2 id="ex_label">Example</h2>
         <div class="combobox combobox-list">
           <div class="group">
             <input id="cb1-input" class="cb_edit" type="text" role="combobox" aria-autocomplete="both"
-              aria-expanded="false" aria-controls="lb1">
-            <button type="button" id="cb1-button" aria-label="Open" tabindex="-1">
+              aria-expanded="false" aria-controls="cb1-listbox">
+            <button type="button" id="cb1-button" aria-label="State" aria-expanded="false" aria-controls="cb1-listbox" tabindex="-1">
```
Comment (mcking65) [inline examples/combobox/combobox-autocomplete-both.html]: The name of the listbox is "States" plural, and this button controls that, so I recommend:

```suggestion
            <button type="button" id="cb1-button" aria-label="States" aria-expanded="false" aria-controls="cb1-listbox" tabindex="-1">
```

File: examples/combobox/combobox-autocomplete-list.html [hunk 2]
```diff
@@ -53,7 +53,7 @@ <h2 id="ex_label">Example</h2>
           <div class="group">
             <input id="cb1-input" class="cb_edit" type="text" role="combobox" aria-autocomplete="list"
               aria-expanded="false" aria-controls="cb1-listbox">
-            <button id="cb1-button" tabindex="-1" aria-label="Open">
+            <button id="cb1-button" tabindex="-1"  aria-label="State" aria-expanded="false" aria-controls="cb1-listbox">
```
Comment (mcking65) [inline examples/combobox/combobox-autocomplete-list.html]: Same recommendation as above:

```suggestion
            <button id="cb1-button" tabindex="-1"  aria-label="States" aria-expanded="false" aria-controls="cb1-listbox">
```

File: examples/combobox/combobox-autocomplete-none.html [hunk 3]
```diff
@@ -53,7 +53,7 @@ <h2 id="ex_label">Example</h2>
                    aria-autocomplete="none"
                    aria-expanded="false"
                    aria-controls="cb1-listbox">
-            <button type="button" id="cb1-button" tabindex="-1" aria-label="Open">
+            <button type="button" id="cb1-button" tabindex="-1"  aria-label="State" aria-expanded="false" aria-controls="cb1-listbox" >
```
Comment (mcking65) [inline examples/combobox/combobox-autocomplete-none.html]: This one controls the "Previous Searches" list, so the aria-label should reflect that.

```suggestion
            <button type="button" id="cb1-button" tabindex="-1"  aria-label="Previous Searches" aria-expanded="false" aria-controls="cb1-listbox" >
```

File: examples/combobox/combobox-autocomplete-both.html [hunk 4]
```diff
@@ -426,6 +426,44 @@ <h3 id="rps_label_listbox">Listbox Popup</h3>
...
+        <thead>
+          <tr>
+            <th scope="col">Role</th>
+            <th scope="col">Attribute</th>
+            <th scope="col">Element</th>
+            <th scope="col">Usage</th>
+          </tr>
+        </thead>
+        <tbody>
+          <tr data-test-id="button-aria-controls">
```
Comment (mcking65) [inline examples/combobox/combobox-autocomplete-both.html]: I think we need a row before this one for aria-label

File: examples/combobox/combobox-autocomplete-list.html [hunk 5]
```diff
@@ -422,8 +422,44 @@ <h3 id="rps_label_listbox">Listbox Popup</h3>
...
+        <thead>
+          <tr>
+            <th scope="col">Role</th>
+            <th scope="col">Attribute</th>
+            <th scope="col">Element</th>
+            <th scope="col">Usage</th>
+          </tr>
+        </thead>
+        <tbody>
+          <tr data-test-id="button-aria-controls">
```
Comment (mcking65) [inline examples/combobox/combobox-autocomplete-list.html]: I think we need a row before this one for aria-label

File: examples/combobox/combobox-autocomplete-none.html [hunk 6]
```diff
@@ -371,8 +371,43 @@ <h3 id="rps_label_listbox">Listbox Popup</h3>
...
+        <thead>
+          <tr>
+            <th scope="col">Role</th>
+            <th scope="col">Attribute</th>
+            <th scope="col">Element</th>
+            <th scope="col">Usage</th>
+          </tr>
+        </thead>
+        <tbody>
+          <tr data-test-id="button-aria-controls">
```
Comment (mcking65) [inline examples/combobox/combobox-autocomplete-none.html]: I think we need a row before this one for aria-label

Comment (jongund) [reply examples/combobox/combobox-autocomplete-none.html]: Fixed

File: examples/combobox/combobox-autocomplete-list.html (hunk 5 above)
Comment (jongund) [reply examples/combobox/combobox-autocomplete-list.html]: Fixed

File: examples/combobox/combobox-autocomplete-both.html (hunk 4 above)
Comment (jongund) [reply examples/combobox/combobox-autocomplete-both.html]: Fixed

File: examples/combobox/combobox-autocomplete-none.html (hunk 3 above)
Comment (jongund) [reply examples/combobox/combobox-autocomplete-none.html]: Fixed

File: examples/combobox/combobox-autocomplete-list.html (hunk 2 above)
Comment (jongund) [reply examples/combobox/combobox-autocomplete-list.html]: Fixed

File: examples/combobox/combobox-autocomplete-both.html (hunk 1 above)
Comment (jongund) [reply examples/combobox/combobox-autocomplete-both.html]: Fixed

**Comment by jongund** [comment]
@mcking65 

I updated the pull request with your suggestions and also added a row for `tabindex=-``.  I also added a section in the keyboard support section saying that the `button` is being removed from tab sequence of the page, but is still important for assistive technologies that use touch events to open the listbox.

**Comment by jongund** [comment]
@mcking65 

I updated this pull request to also address issue #1331 high contrast mode.

**Comment by zcorpan** [comment]
Visual design review: the relevant change is about Windows high contrast, reviewed in https://github.com/w3c/aria-practices/issues/1331#issuecomment-594011492

File: examples/combobox/combobox-autocomplete-list.html [hunk 7]
```diff
@@ -53,10 +53,11 @@ <h2 id="ex_label">Example</h2>
           <div class="group">
             <input id="cb1-input" class="cb_edit" type="text" role="combobox" aria-autocomplete="list"
               aria-expanded="false" aria-controls="cb1-listbox">
-            <button id="cb1-button" tabindex="-1" aria-label="Open">
+            <button id="cb1-button" tabindex="-1"  aria-label="States" aria-expanded="false" aria-controls="cb1-listbox">
```
Comment (zcorpan) [inline examples/combobox/combobox-autocomplete-list.html]:
```suggestion
            <button id="cb1-button" tabindex="-1" aria-label="States" aria-expanded="false" aria-controls="cb1-listbox">
```

File: test/tests/combobox_autocomplete-both.js [hunk 8]
```diff
@@ -149,6 +149,54 @@ ariaTest('"aria-selected" attribute on options element', exampleFile, 'option-ar
...
+    .findElement(By.css(ex.textboxSelector))
+    .getAttribute('aria-controls');
+
+  const popupElement = await t.context.session
+    .findElement(By.id('ex1'))
+    .findElement(By.id(popupId));
+
+  t.false(
+    await popupElement.isDisplayed(),
+    'Popup element should not be displayed when \'aria-expanded\' is false\''
```
Comment (zcorpan) [inline test/tests/combobox_autocomplete-both.js]:
```suggestion
    'Popup element should not be displayed when \'aria-expanded\' is \'false\''
```

File: test/tests/combobox_autocomplete-both.js [hunk 9]
```diff
@@ -149,6 +149,54 @@ ariaTest('"aria-selected" attribute on options element', exampleFile, 'option-ar
...
+
+  t.is(
+    await button.getAttribute('aria-expanded'),
+    'true',
+    'button element should have attribute "aria-expand" set to true after typing.'
+  );
+
+  t.true(
+    await popupElement.isDisplayed(),
+    'Popup element should be displayed when \'aria-expanded\' is true\''
```
Comment (zcorpan) [inline test/tests/combobox_autocomplete-both.js]:
```suggestion
    'Popup element should be displayed when \'aria-expanded\' is \'true\''
```

File: test/tests/combobox_autocomplete-none.js [hunk 10]
```diff
@@ -154,6 +155,75 @@ ariaTest('"aria-selected" attribute on options element', exampleFile, 'option-ar
...
+    .findElement(By.css(ex.textboxSelector))
+    .getAttribute('aria-controls');
+
+  const popupElement = await t.context.session
+    .findElement(By.id('ex1'))
+    .findElement(By.id(popupId));
+
+  t.false(
+    await popupElement.isDisplayed(),
+    'Popup element should not be displayed when \'aria-expanded\' is false\''
```
Comment (zcorpan) [inline test/tests/combobox_autocomplete-none.js]:
```suggestion
    'Popup element should not be displayed when \'aria-expanded\' is \'false\''
```

File: test/tests/combobox_autocomplete-none.js [hunk 11]
```diff
@@ -154,6 +155,75 @@ ariaTest('"aria-selected" attribute on options element', exampleFile, 'option-ar
...
+
+  t.is(
+    await button.getAttribute('aria-expanded'),
+    'true',
+    'button element should have attribute "aria-expand" set to true after typing.'
+  );
+
+  t.true(
+    await popupElement.isDisplayed(),
+    'Popup element should be displayed when \'aria-expanded\' is true\''
```
Comment (zcorpan) [inline test/tests/combobox_autocomplete-none.js]:
```suggestion
    'Popup element should be displayed when \'aria-expanded\' is \'true\''
```

File: examples/combobox/combobox-autocomplete-none.html [hunk 12]
```diff
@@ -53,7 +53,7 @@ <h2 id="ex_label">Example</h2>
                    aria-autocomplete="none"
                    aria-expanded="false"
                    aria-controls="cb1-listbox">
-            <button type="button" id="cb1-button" tabindex="-1"  aria-label="State" aria-expanded="false" aria-controls="cb1-listbox" >
+            <button type="button" id="cb1-button" tabindex="-1"  aria-label="Previous Searches" aria-expanded="false" aria-controls="cb1-listbox" >
```
Comment (zcorpan) [inline examples/combobox/combobox-autocomplete-none.html]:
```suggestion
            <button type="button" id="cb1-button" tabindex="-1" aria-label="Previous Searches" aria-expanded="false" aria-controls="cb1-listbox" >
```

**Comment by zcorpan** [comment]
Maybe https://github.com/w3c/aria-practices/pull/1343 should be merged first, and this be rebased on top and remove the `t.plan(1)`s.

**Comment by zcorpan** [comment]
> Maybe #1343 should be merged first, and this be rebased on top and remove the `t.plan(1)`s.

Never mind; I think that PR should add a lint check, which means other PRs like this one don't need to be blocked on it.

**Comment by zcorpan** [comment]
`npm run regression-report` gives

```
combobox/combobox-autocomplete-both.html:
    combobox-id
combobox/combobox-autocomplete-list.html:
    combobox-id
combobox/combobox-autocomplete-none.html:
    combobox-id
```

I think those failures were introduced by #1276. I can report a new issue about this.

**Comment by zcorpan** [comment]
Actually I can't merge, I'm not authorized to push to protected branches.

**Pull request merged**